File: Python/Concurrency/ReadWriteLockDemo.py
import threading
import time
from concurrent.futures import ThreadPoolExecutor


class SharedCounter:
    def __init__(self) -> None:
        self._value = 0
        self._write_lock = threading.Lock()
        self._readers_lock = threading.Lock()
        self._readers = 0

    def read(self) -> int:
        with self._readers_lock:
            self._readers += 1
            if self._readers == 1:
                self._write_lock.acquire()
        try:
            print(f"{threading.current_thread().name} reading: {self._value}")
            self._sleep(0.1)  # simulate read taking some time
            return self._value
        finally:
            with self._readers_lock:
                self._readers -= 1
                if self._readers == 0:
                    self._write_lock.release()

    def write(self, new_value: int) -> None:
        with self._write_lock:
            print(f"{threading.current_thread().name} WRITING: {new_value}")
            self._sleep(0.1)
            self._value = new_value

    @staticmethod
    def _sleep(seconds: float) -> None:
        time.sleep(seconds)


def main() -> None:
    counter = SharedCounter()

    with ThreadPoolExecutor(max_workers=5) as pool:
        # 4 readers — watch their timestamps overlap
        for _ in range(4):
            pool.submit(counter.read)

        time.sleep(0.02)  # let readers start first

        # 1 writer — watch it wait for ALL readers to finish, then block everyone else
        pool.submit(counter.write, 99)


if __name__ == "__main__":
    main()

File: Java/Concurrency/ReadWriteLockDemo.java
package Concurrency;
import java.util.concurrent.locks.*;
import java.util.concurrent.*;

public class ReadWriteLockDemo {

    static class SharedCounter {
        private int value = 0;
        private final ReadWriteLock rwLock = new ReentrantReadWriteLock();

        public int read() {
            rwLock.readLock().lock();
            try {
                System.out.println(Thread.currentThread().getName() + " reading: " + value);
                sleep(100);  // simulate read taking some time
                return value;
            } finally {
                rwLock.readLock().unlock();
            }
        }

        public void write(int newValue) {
            rwLock.writeLock().lock();
            try {
                System.out.println(Thread.currentThread().getName() + " WRITING: " + newValue);
                sleep(100);
                value = newValue;
            } finally {
                rwLock.writeLock().unlock();
            }
        }

        private void sleep(long ms) {
            try { Thread.sleep(ms); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        SharedCounter counter = new SharedCounter();
        ExecutorService pool = Executors.newFixedThreadPool(5);

        // 4 readers — watch their timestamps overlap
        for (int i = 0; i < 4; i++) {
            pool.submit(counter::read);
        }

        Thread.sleep(20); // let readers start first

        // 1 writer — watch it wait for ALL readers to finish, then block everyone else
        pool.submit(() -> counter.write(99));

        pool.shutdown();
        pool.awaitTermination(5, TimeUnit.SECONDS);
    }
}